Revert Homematic IP Cloud unique ID changes

hahn-th opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Revert latest change in generation of unique-ids. 
Therefore changes regarding the Sensor Hmip-ESI are necessary.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #121130 
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

hahn-th committed:
Fix unique-id generation

hahn-th committed:
Add missing tests

hahn-th commented:
> Shouldn't this PR in this case clean up the issue / potential duplicates it caused?

The title of the issue is somewhat misleading. In fact the entities got a new unique_id because of the change in the unique_id property. That's why new entities where added and the old ones where deprecated.
![image](https://github.com/home-assistant/core/assets/15319212/b751f422-f67e-4af2-a5fa-24212a5b9562)

I had a look at the code and found a solution that makes everything work again. 
The unique_id property in generic_entity class works like before now. The unique_id is generated out of the class name. So i gave every of the HmIP-ESI sensors an own class instead of using EntityDescriptions.

The unitests in test_devices.py::test_hmip_load_all_supported_devices loads all devices without duplicates.

frenck reviewed (approved):
Thanks, @hahn-th 👍 

../Frenck

glaskopf commented:
Thx for the quick fix - waiting now for 2024.7.1 update

NikoGrub commented:
My entities are now renamed again after the 2024.7.1 update.
I just renamed all entities again manually and fixed the automations.

Just for understanding;

Is the problem now gone completely, or will it be the same procedure after the next update?

Thank you for your work!

hahn-th commented:
> My entities are now renamed again after the 2024.7.1 update. I just renamed all entities again manually and fixed the automations.
> 
> Just for understanding;
> 
> Is the problem now gone completely, or will it be the same procedure after the next update?
> 
> Thank you for your work!

No more renaming. They are named like before the 2024.7.0 update.

glaskopf commented:
I can confirm this.
 Everything is now working as expected. 
The addition of the energy counter and the usage in the energy dashboard  is really  a great extension. Thank you again very much !
